docs/design/interpreter/compiled-code-interop.md

```diff
@@ -0,0 +1,69 @@
+
+##### Interpreter exit
+
+In order to support the pinvoke or compiled code call paths, we could use the same approach as with the native architecture. The only difference is that we won't be able to have a generic path, but for this transition it should rarely be problematic  since we have a clear picture of the code that we would need to invoke into and signatures are typically reused. When the application is aot compiled, we will include a compiled wrapper for every signature of a compiled method as well as for every pinvoke signature. The wrapper will receive the target pointer to call and the address of the interpreter stack where the arguments are present. On mono these wrappers are written in C with dynamically generated code during app compilation time. I think it makes more sense to include them as compiled IL wrappers which should allow for code reuse with the native architecture approach and also for the invocation path to be as fast as possible.
+
+##### Interpreter entry
+
+Given we can't dynamically generate thunks that can be invoked, in order to generate an interpreter entry point we could reuse the functionality of fat pointers that is already used with native aot. Fat pointers can point to additional data, if a bit is set, rather than being actual function pointers. Instead of calling the pointer directly, the calling code will check for the most significant bit. If it is set, it will instead dereference the pointer and obtain the real function pointer together with the additional argument that is passed to the call. For the purpose of entering the interpreter, we would generate a fat pointer that has the target destination as a compiled IL wrapper for the signature in question together with the interpreter method pointer that is passed. The wrapper will obtain the pointer to the interpreter stack, will move all arguments there and call into the C++ interpreter path, passing the method and the address on the stack where the arguments have been written. When aot compiling an assembly, we would need to consider every single call as a potential entry to the interpreter and, if we deem it possible, we would generate an interpreter entry wrapper for the call signature in question.
```

Comment:
I believe fat pointers on Wasm use the MSB, but on non-Wsm, the LSB is used (beacuse Wsam function pointers are indicies)